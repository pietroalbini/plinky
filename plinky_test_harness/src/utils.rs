use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
