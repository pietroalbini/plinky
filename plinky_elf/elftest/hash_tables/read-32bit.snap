---
source: plinky_test_harness/src/utils.rs
---
reading ELF exited with exit status: 0

=== stdout ===
section .hash#1 (address: 0x114)
 │
 │  Hash table for .dynsym#2
 │
 │    Content:
 │  ╭───────────┬───────────────────╮
 │  │ Bucket ID │ Symbols in bucket │
 │  ├───────────┼───────────────────┤
 │  │ 0         │ c#9               │
 │  │           │ i#8               │
 │  │           │ l#6               │
 │  │           │ f#4               │
 │  ├───────────┼───────────────────┤
 │  │ 1         │ m#14              │
 │  │           │ a#12              │
 │  │           │ d#10              │
 │  │           │ g#5               │
 │  │           │ j#2               │
 │  ├───────────┼───────────────────┤
 │  │ 2         │ e#13              │
 │  │           │ h#11              │
 │  │           │ k#7               │
 │  │           │ n#3               │
 │  │           │ b#1               │
 │  ╰───────────┴───────────────────╯
 ┴

section .dynsym#2 (address: 0x164)
 │
 │    Dynamic symbol table:
 │  ╭──────┬─────────┬──────────┬────────────┬────────────┬────────┬──────╮
 │  │ Name │ Binding │ Type     │ Visibility │ Definition │ Value  │ Size │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ #0   │ Local   │ -        │ Default    │ Undefined  │ 0x0    │ 0x0  │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ b#1  │ Global  │ Function │ Default    │ .text#4    │ 0x1010 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ j#2  │ Global  │ Function │ Default    │ .text#4    │ 0x1090 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ n#3  │ Global  │ Function │ Default    │ .text#4    │ 0x10d0 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ f#4  │ Global  │ Function │ Default    │ .text#4    │ 0x1050 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ g#5  │ Global  │ Function │ Default    │ .text#4    │ 0x1060 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ l#6  │ Global  │ Function │ Default    │ .text#4    │ 0x10b0 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ k#7  │ Global  │ Function │ Default    │ .text#4    │ 0x10a0 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ i#8  │ Global  │ Function │ Default    │ .text#4    │ 0x1080 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ c#9  │ Global  │ Function │ Default    │ .text#4    │ 0x1020 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ d#10 │ Global  │ Function │ Default    │ .text#4    │ 0x1030 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ h#11 │ Global  │ Function │ Default    │ .text#4    │ 0x1070 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ a#12 │ Global  │ Function │ Default    │ .text#4    │ 0x1000 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ e#13 │ Global  │ Function │ Default    │ .text#4    │ 0x1040 │ 0x10 │
 │  ├──────┼─────────┼──────────┼────────────┼────────────┼────────┼──────┤
 │  │ m#14 │ Global  │ Function │ Default    │ .text#4    │ 0x10c0 │ 0x10 │
 │  ╰──────┴─────────┴──────────┴────────────┴────────────┴────────┴──────╯
 ┴

no stderr present
